mirgecom/profiling.py

```diff
@@ -0,0 +1,208 @@
+
+        print(format_str.format('='*20, '='*6, '='*8, '='*8, '='*8, '='*8, '='*8,
+            '='*8, '='*8, '='*8, '='*8, '='*8))
+
+    def call_loopy(self, program, **kwargs) -> dict:
+        program = self.transform_loopy_program(program)
+        assert program.options.return_dict
+        assert program.options.no_numpy
+
+        if self.profiling_enabled:
+            args_tuple = ()
+
+            for key, value in kwargs.items():
+                if hasattr(value, 'shape'):
+                    args_tuple = args_tuple + (key, value.shape)
+                else:
+                    args_tuple = args_tuple + (key, value)
+
+            # Check if we need to get the invoker code to generate integer arguments
+            # N.B.: The invoker code might be different for the same program with
+            # different kwargs
+            if (program not in self.kernel_stats
+              or args_tuple not in self.kernel_stats[program]):
+                executor = program.target.get_kernel_executor(program, self.queue)
+                info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+                kernel = executor.get_typed_and_scheduled_kernel(
+                    executor.arg_to_dtype_set(kwargs))
+
+                data = info.implemented_data_info
+
+                # Generate the wrapper code
+                wrapper = executor.get_wrapper_generator()
+
+                gen = PythonCodeGenerator()
+
+                wrapper.generate_integer_arg_finding_from_shapes(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_offsets(gen, kernel, data)
+                wrapper.generate_integer_arg_finding_from_strides(gen, kernel, data)
+
+                types = {}
+                param_dict = {}
+
+                for key, value in kwargs.items():
+                    if hasattr(value, 'dtype') and not value.dtype == object:
+                        types[key] = value.dtype
+                    param_dict[key] = value
+
+                for key, value in kernel.arg_dict.items():
+                    if key not in param_dict:
+                        param_dict[key] = None
+
+                for d in data:
+                    if d.name not in param_dict:
+                        param_dict[d.name] = None
+
+                # Run the wrapper code, save argument values in param_dict
+                exec(gen.get(), param_dict)
+
+                # Get flops/memory statistics
+                kernel.options.ignore_boostable_into = True
+                kernel = lp.add_and_infer_dtypes(kernel, types)
+                op_map = lp.get_op_map(kernel, subgroup_size='guess')
+                mem_map = lp.get_mem_access_map(kernel, subgroup_size='guess')
+
+                f32op = op_map.filter_by(dtype=[np.float32]).eval_and_sum(param_dict)
+                f64op = op_map.filter_by(dtype=[np.float64]).eval_and_sum(param_dict)
```

Comment:
Why specialize to those two?

Reply:
Changed to `np.float`.

Reply:
Float is an alias for `float64`. Not better.

Reply:
Ok, what do you suggest? Add a line for `np.float16`?

Reply:
On second thought, I think this is OK. Code review puts me in a pretty critical mood.

Reply:
Ok, I simplified the code here a bit nevertheless.